cpp/celeborn/protocol/ControlMessages.cpp

```
@@ -0,0 +1,187 @@
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "celeborn/protocol/ControlMessages.h"
+#include "celeborn/utils/CelebornUtils.h"
+
+namespace celeborn {
+namespace protocol {
+TransportMessage GetReducerFileGroup::toTransportMessage() const {
+  MessageType type = GET_REDUCER_FILE_GROUP;
+  PbGetReducerFileGroup pb;
+  pb.set_shuffleid(shuffleId);
+  std::string payload = pb.SerializeAsString();
+  return TransportMessage(type, std::move(payload));
+}
+
+std::unique_ptr<GetReducerFileGroupResponse>
+GetReducerFileGroupResponse::fromTransportMessage(
+    const TransportMessage& transportMessage) {
+  CELEBORN_CHECK(
+      transportMessage.type() == GET_REDUCER_FILE_GROUP_RESPONSE,
+      "transportMessageType mismatch");
+  auto payload = transportMessage.payload();
+  auto pbGetReducerFileGroupResponse =
+      utils::parseProto<PbGetReducerFileGroupResponse>(
+          reinterpret_cast<const uint8_t*>(payload.c_str()), payload.size());
+  auto response = std::make_unique<GetReducerFileGroupResponse>();
+  response->status = toStatusCode(pbGetReducerFileGroupResponse->status());
+  auto fileGroups = pbGetReducerFileGroupResponse->filegroups();
+  for (auto& kv : fileGroups) {
+    auto& fileGroup = response->fileGroups[kv.first];
+    // Legacy mode: the locations are valid, so use it.
```

shall we need support legacy mode?

---

In java end,  the legacy mode is kept for backward compatibility. I think keeping legacy mode in CppClient does no harm, or should we simply throw an error here if the legacy mode is detected?
Any suggestion is welcome.

---

Native SDK will be shipped in 0.6 or 0.7. Should use new format as default.

---

Refactored as requested.
Any suggestion is welcome.